File: android/src/main/kotlin/com/winkk/flutter_native_mutex/FlutterNativeMutexPlugin.kt
package com.winkk.flutter_native_mutex

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result
import java.util.ArrayDeque

class FlutterNativeMutexPlugin: FlutterPlugin, MethodCallHandler {
  companion object {
    // An entry represents the current owner; its queue contains only waiting callers.
    // Registration, handoff, and removal must use the same monitor.
    private val sharedMutexMap = mutableMapOf<String, ArrayDeque<Result>>()
  }

  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "winkk/flutter_native_mutex")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(call: MethodCall, result: Result) {
    when (call.method) {
      "lock" -> {
        val globalKey = call.argument<String>("globalKey")
        if (globalKey != null) {
          val acquired = synchronized(sharedMutexMap) {
            val waiters = sharedMutexMap[globalKey]
            if (waiters == null) {
              sharedMutexMap[globalKey] = ArrayDeque()
              true
            } else {
              waiters.addLast(result)
              false
            }
          }
          if (acquired) result.success(null)
        } else {
          result.error("Invalid argument", "globalKey is required", null)
        }
      }
      "unlock" -> {
        val globalKey = call.argument<String>("globalKey")
        if (globalKey != null) {
          val (wasLocked, next) = synchronized(sharedMutexMap) {
            val waiters = sharedMutexMap[globalKey]
            if (waiters == null) {
              false to null
            } else {
              val next = waiters.pollFirst()
              if (next == null) sharedMutexMap.remove(globalKey)
              true to next
            }
          }
          // Invoke callbacks outside the monitor so reentrant channel activity is safe.
          if (wasLocked) {
            next?.success(null)
            result.success(null)
          } else {
            result.error("Invalid argument", "mutex must be locked first", null)
          }
        } else {
          result.error("Invalid argument", "globalKey is required", null)
        }
      }
      else -> result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}

File: ios/flutter_native_mutex/Sources/flutter_native_mutex/FlutterNativeMutexPlugin.swift
import Flutter
import UIKit

public class FlutterNativeMutexPlugin: NSObject, FlutterPlugin {
  // Presence marks an owner. Queued callbacks are handed the lock without blocking a thread.
  private static var sharedMutexMap = [String: [FlutterResult]]()
  private static let accessQueue = DispatchQueue(label: "com.winkk.nativeMutex.accessQueue")
  
  public static func register(with registrar: FlutterPluginRegistrar) {
    let channel = FlutterMethodChannel(name: "winkk/flutter_native_mutex", binaryMessenger: registrar.messenger())
    let instance = FlutterNativeMutexPlugin()
    registrar.addMethodCallDelegate(instance, channel: channel)
  }

  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "lock":
      guard let args = call.arguments as? [String: Any],
            let globalKey = args["globalKey"] as? String else {
          result(FlutterError(code: "Invalid argument", message: "globalKey is required", details: nil))
          return
      }
      let acquired = FlutterNativeMutexPlugin.accessQueue.sync {
        if FlutterNativeMutexPlugin.sharedMutexMap[globalKey] != nil {
          FlutterNativeMutexPlugin.sharedMutexMap[globalKey]!.append(result)
          return false
        }
        FlutterNativeMutexPlugin.sharedMutexMap[globalKey] = []
        return true
      }
      if acquired { result(nil) }
    case "unlock":
      guard let args = call.arguments as? [String: Any],
            let globalKey = args["globalKey"] as? String else {
          result(FlutterError(code: "Invalid argument", message: "globalKey is required", details: nil))
          return
      }
      let (wasLocked, next): (Bool, FlutterResult?) = FlutterNativeMutexPlugin.accessQueue.sync {
        guard var waiters = FlutterNativeMutexPlugin.sharedMutexMap[globalKey] else {
          return (false, nil)
        }
        if waiters.isEmpty {
          FlutterNativeMutexPlugin.sharedMutexMap.removeValue(forKey: globalKey)
          return (true, nil)
        }
        let next = waiters.removeFirst()
        FlutterNativeMutexPlugin.sharedMutexMap[globalKey] = waiters
        return (true, next)
      }
      // Callbacks may reenter the plugin, so never invoke them on accessQueue.
      if wasLocked {
        next?(nil)
        result(nil)
      } else {
        result(FlutterError(code: "Invalid argument", message: "mutex must be locked first", details: nil))
      }
    default:
      result(FlutterMethodNotImplemented)
    }
  }
}
